### DISCUSSION TITLE
Changing parameters of a viewer for different positions of scan

### DISCUSSION THREAD

**Discussion by rgeneaux** [PyMoDAQ/PyMoDAQ#95] [Q&A] [answered]
Hi everyone,

I've ran into an interesting issue the other day: we needed to do a very simple scan with one actuator (translation) alternating between position A and B, and measuring a spectrum (1D viewer) at each position. 
But for it to work, we would need to integrate 100 ms at position A, and 3s at position B! (one is a direct beam, the one is the beam absorbed by a sample)

Do you have an idea of how this could be done? I thought of using a Mock Actuator that would "fake move" together with the translation and could control the exposure time of the viewer, but I didn't figure out how to make the daq_move_stage and daq_viewer communicate with each other.

Thanks!
Romain

**Comment by LMSC-NTappy** [comment, marked as answer]
Hi Romain!

That sounds a bit tricky, I never had to do something like this but here are my thoughts on this:

If I understood correctly you have a DAQ_NDViewer object with an "integration time" float setting that you want to change accordingly to the position of an A/B DAQ_Move. 

What I would do is create a new DAQ_Move plugin that relies on the same wrapper as your DAQ_NDViewer, that sets the integration time parameter. Then in your preset you can initialize DAQ_Move as master and DAQ_NDViewer as slave. Then you can use it as an actuator to update the setting of your DAQ_Viewer. 

I used this in a plugin that controls a Keithley ampèremetre + voltage source, you can see it here:
https://github.com/LMSC-NTappy/pymodaq_plugins_keithley

It worked for me when I needed to scan the source voltage while measuring current, so I had a single actuator scan there. To sync this with the binary actuator I would then use the tabular scan as explained here:
http://pymodaq.cnrs.fr/en/latest/usage/modules/Utility_Modules.html#tabular

It's probably a bit painful to define values manually but I guess you can create a script that generate the .txt file with the scan positions.

I hope this will work out for you. Please let us know!

Cheers and have a happy 2023!

Nicolas

**Reply by seb5g** [reply]
I think that's the right direction knowing how the DAQ_Scan works. For the tabular, there is a contextual menu when you right click on the tabular settings. That allow you to load a text file containing your values. Another fancier approach is to build a small interface that lets you specify your two positions and your two integrations, calculate all the values for the tabular and then plug this appli on to the entry point of the tabular scan. In the pymodaq examples, you'll find the NonLinearScanner that do something very similar

**Reply by rgeneaux** [reply]
Worked like a charm! 

If it is of use to anyone in the future, here how we implemented this:
- Modify the `DAQ_2DViewer_Camera` so that `self.controller` has access to the exposure time. It gets an attribute `self.controller.current_exposure_time` that gets updated when the exposure changes.
 
- Create `DAQ_Move_CameraExposureController` which is to be used as a slave of the camera in the dashboard, so that it has access to the camera's self.controller. It will return an error message when not declared as a slave. It's only job is through self.get_actuator_value(), which returns `self.controller.current_exposure_time`, and self.move_abs(exp) which updates `self.controller.current_exposure_time`. 

- Everytime the camera starts an acquisition, it will check what is  `self.controller.current_exposure_time` (in case it was changed externally by CameraExposureController) and update its exposure if necessary.

- Then the Tabular scan can easily be set up - indeed the easiest was to create the scan points in notepad, and right click the tabular to use the load from .txt functionality.

Thanks for the help!!
